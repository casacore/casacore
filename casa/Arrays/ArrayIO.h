//# ArrayIO.h: text output and binary IO for an array of any dimensionality.
//# Copyright (C) 1993,1994,1995,1997,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_ARRAYIO_H
#define CASA_ARRAYIO_H

//# Includes
#include <casa/aips.h>
#include <casa/BasicSL/String.h>

//# Forward declarations
#include <casa/iosfwd.h>

namespace casa { //# NAMESPACE CASA - BEGIN

class AipsIO;
class LogIO;
class IPosition;
template<class T> class Array;
template<class T> class Matrix;
template<class T> class Vector;
template<class T> class Cube;
template<class T> class Block;

// <summary>
//    Input/output operators for Arrays.
// </summary>

// <use visibility=export>

// <reviewed reviewer="Paul Shannon" date="1995/02/21" tests="" demos="">
// This header was reviewed and revised with the goal of making it an
// example for those writing global function header files.
// </reviewed>

// <prerequisite>
//   <li> <linkto class=Array>Array</linkto>
//   <li> ostream
//   <li> <linkto class=AipsIO>AipsIO</linkto>
// </prerequisite>

// <etymology>
//  ArrayIO is simply the conventional shorthand for "array input/output".
// </etymology>

// <synopsis> 
// These global functions provide easy input and output of (possibly)
// large and (possibly) multi-dimensional arrays.  Iteration through
// entire arrays is done behind the scenes, with no effort required
// of the client programmer.  
// These functions are global, rather than member functions of the
// Array class, because of the well-known C++ requirement that the first
// argument to an operator function (as it is declared) is the 
// left operand when the function is called.
// </synopsis>


// <example>
// <srcblock>
// IPosition shape (3,10,10,3);
// Array <Float> array (shape);
// // ...initialize and manipulate the array...
// cout << "result: " << array;
// </srcblock>
//
// <motivation>
// Effortless input/output is clearly a big win.
// </motivation>
//
// <todo asof="1997/01/15">
// </todo>

// <linkfrom anchor="Array IO" classes="Array Vector Matrix Cube">
//    <here>Array IO</here> -- Input/output operators for Arrays.
// </linkfrom>
//
// <group name="Array IO">


// Write out an ascii representation of an array of any dimensionality.
// Arrays of dimensionality 3 or greater are written out vector by vector,
// preceeded by the position of the start of the vector. If the origin of
// the array isn't zero it is printed. The shape of the array is always
// printed.

template<class T> ostream &operator << (ostream &, const Array<T> &);

// Write a formatted copy of the array to the LogIO output object. Merely calls
// the ostream operator<< in turn.
template<class T> LogIO &operator<<(LogIO &os, const Array<T> &a);

// Read an ascii representation of an array. All types with an <src><<</src>
// operator can be handled. The basic format of the input should be:
// <srcblock>
//	[element element element ....] 
// </srcblock>
// Elements are separated by whitespace, or a comma, optionally surrounded
// by white space. <br>
// <note role=warning> Some input routines read fields between blank spaces. This
// is (at the moment) especially true for Quantities and Strings.
// In those cases
// the separator should be blank (or a comma following a blank), and the
// end ']' should have a blank in front.
// A crude fix for String arrays having separators <src>,</src> and <src>]</src>
// without blanks preceding has been made; but slows routines down </note>
// The default input is a vector of unspecified length. The input shape
// can be changed by pre-pending the input with:
// <srcblock>
//	{[shape]}
// </srcblock>
// where shape is an unsigned integer vector. The shape will be used to check
// the input length; and, depending on the possibility, to resize/reshape the
// result. However, reshaping of e.g. a Vector to a Matrix cannot be done, and
// the result will stay in the form asked.<br>
// Input order is row major, however by preceding the input with:
// <srcblock>
//	{T[shape]}	
// </srcblock>
// the order will be reversed.<br>
// Reshaping of the Array provided will depend on the type of Array and its
// state. If a general Array, the shape will be
// as defined by user. If fixed Array (e.g. Matrix, Vector, Cube) the number
// of dimesnsions will be kept. If the user specified more dimensions
// then supported (e.g. 3 for Matrix), the last dimesions will be collapsed.
// If less dimensions are specified, the missing ones will be set to 1.
// will be kept.<br>
// The read() version can be used to force a shape (ip), or an input
// transpose (it) (which can be undone by the user specifying transpose).
//
// <group>

template<class T> istream &operator >> (istream &s, Array<T> &x);
template<class T> Bool read(istream &s, Array<T> &x,
			    const IPosition *ip=0, Bool it=False);
// </group>

// General read support function for matrices.
// In principle these functions will not be
// used by general user, but could be. They can be used by Array type
// classes (like Slice, Lattice) to do the work of comparable input
// functions as the one for Arrays.
// In these functions p is the shape
// of the returned Block x. This shape is either deduced from the user
// specification; made equal to (1, nelements) if no user shape is
// given; is set to ip if specified. The function will return False (and
// p = (0)) in the case of an invalid input element; a number of elements
// input not equal to ip (if specified); the shape given by user as input
// does not conform to ip (if given) or the number of elements input.<br>
// trans will be True if transpose asked by user; or if forced by it.


template<class T> Bool readArrayBlock(istream &s, Bool &trans,
				      IPosition &p,
				      Block<T> &x,
				      const IPosition *ip=0, Bool it=False);


// Read or write a binary representation of an Array to a file. Very
// useful for saving arrays and restoring them later.
// <br>The putArray function is put in for forwards compatibility
// of images (so new images can be read with old release).
//
// <group>

template<class T> AipsIO &operator<< (AipsIO &, const Array<T> &);
template<class T> void putArray (AipsIO &, const Array<T> &, const Char* name);
template<class T> AipsIO &operator>> (AipsIO &, Array<T> &);

// </group>

// </group>




// <summary>
// Global functions to read/write binary arrays from/to a file.
// </summary>

// <use visibility=export>

// <reviewed reviewer="Gareth Hunt" date="95Mar31" tests="" demos="">
// </reviewed>

// <synopsis>
// These global functions provide disk read/write functions for an Array of
// binary numbers.  The write operation is useful, for example, to dump an
// image in binary form to disk so that it can be displayed with an external
// utility such as SAOimage.
// </synopsis>

// <example>
// <srcblock>
//    Matrix<Float> picture(256, 256); picture = 0.0;
//    String fileName="picture.data";
//
//    // operations to populate picture
//    // ...
//
//    write_array (picture, fileName);
// </srcblock>
// </example>

// <todo asof="">
//   <li> These functions should eventually be replaced with something
//        more sophisticated.
// </todo>

// <linkfrom anchor="Array binary IO" classes="Array Vector Matrix Cube">
//    <here>Array binary IO</here> -- Simple binary input/output for Arrays.
// </linkfrom>

// <group name=Array binary IO>

// Write the values of an array in binary format into a file with
// the given name.
// The values are stored in local format, thus are not converted
// to a canonical format as
// <linkto class="AipsIO:description">AipsIO</linkto>
// does.
// <note role=warning>
// This function is only suitable for built-in data types.
// </note>
// <group>
template <class T>
void write_array (const Array<T>& the_array, const String& fileName);

template <class T>
inline void write_array (const Array<T>& the_array, const Char* fileName)
    { write_array (the_array, String(fileName)); }
// </group>

// Read the values of an array in binary format from a file with
// the given name.
// The number of values read is the size of the Array, thus the file
// should at least contain that number of values.
// <note role=warning>
// This function is only suitable for built-in data types.
// </note>
// <group>
template <class T>
void read_array (Array<T>& the_array, const String& fileName);

template <class T>
inline void read_array (Array<T>& the_array, const Char* fileName)
    { read_array (the_array, String(fileName)); }
// </group>

// </group>



// <summary>
// Global functions for Matrix/Vector input/output using ASCII format.
// </summary>

// <use visibility=export>

// <prerequisite>
//   <li> <linkto class=Matrix>Matrix</linkto>
//   <li> <linkto class=Vector>Vector</linkto>
// </prerequisite>

// <synopsis>
// These global functions support file I/O between ASCII files and
// Matrices or Vectors.
// </synopsis>

// <example>
// <srcblock>
//    Matrix<Float> picture(256, 256); picture = 0.0;
//    String fileName="picture.data";
//
//    // operations to populate picture
//    // ...
//
//    writeAsciiMatrix (picture, fileName);
// </srcblock>
// </example>

// <linkfrom anchor="Array Ascii IO" classes="Vector Matrix">
//    <here>Array Ascii IO</here> -- Simple Ascii input/output for Arrays.
// </linkfrom>

// <group name=Array Ascii IO>

// These routines read and write a Matrix of data.  The first line of
// input will be examined to determine the number of columns in the matrix.
// The maximum number of columns provided for is 100.  Each item may be up 
// to 50 characters long.
//
// Each item must be separated from others by one (or more) blank column.
// The "line" may be up to 1024 characters long.  Each subsequent line must
// contain the SAME number of items as the first line but may be any length
// (up to 1024 characters).
//
// The matrix need NOT be square.  
//
// The matrix should be declared but NOT dimensioned in the calling program.

// <group>
template <class T>
void readAsciiMatrix (Matrix<T>& mat, const Char* fileName);

template <class T>
void writeAsciiMatrix (const Matrix<T>& mat, const Char* fileName);
// </group>


// These two functions read and write a Vector of data.  The input
// may be arranged in any format (i.e. It may be recorded as one value per
// line or it may be recorded with all values on a single line).
// Values must be separated by whitespace.

// <group>
template <class T>
void readAsciiVector (Vector<T>& vec, const Char* fileName);

template <class T>
void writeAsciiVector (const Vector<T>& vec, const Char* fileName);
// </group>

// </group>



} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <casa/Arrays/ArrayIO.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
