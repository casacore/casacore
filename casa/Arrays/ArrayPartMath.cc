//# ArrayPartMath.h: mathematics done on an array parts.
//# Copyright (C) 1993,1994,1995,1996,1998,1999,2001,2003
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: ArrayPartMath.cc 21262 2012-09-07 12:38:36Z gervandiepen $

#include <casacore/casa/Arrays/ArrayPartMath.h>

namespace casacore {

  void fillBoxedShape (const IPosition& shape, const IPosition& boxSize,
                       IPosition& fullBoxSize, IPosition& resultShape)
  {
    uInt ndim = shape.size();
    // Set missing axes to 1.
    fullBoxSize.resize (ndim);
    fullBoxSize = 1;
    for (uInt i=0; i<min(ndim,boxSize.size()); ++i) {
      // Set unspecified axes to full length.
      if (boxSize[i] <= 0  ||  boxSize[i] > shape[i]) {
        fullBoxSize[i] = shape[i];
      } else {
        fullBoxSize[i] = boxSize[i];
      }
    }
    // Determine the output shape.
    resultShape.resize (ndim);
    for (uInt i=0; i<ndim; ++i) {
      resultShape[i] = (shape[i] + fullBoxSize[i] - 1) / fullBoxSize[i];
    }
  }

  Bool fillSlidingShape (const IPosition& shape, const IPosition& halfBoxSize,
                         IPosition& boxEnd, IPosition& resultShape)
  {
    uInt ndim = shape.size();
    // Set full box end (is size-1) and resize/fill as needed.
    boxEnd.resize (halfBoxSize.size());
    boxEnd = 2*halfBoxSize;
    if (boxEnd.size() != ndim) {
      uInt sz = boxEnd.size();
      boxEnd.resize (ndim);
      for (uInt i=sz; i<boxEnd.size(); ++i) {
        boxEnd[i] = 0;
      }
    }
    // Determine the output shape. See if anything has to be done.
    Bool empty = False;
    resultShape.resize (shape.size());
    for (uInt i=0; i<ndim; ++i) {
      resultShape[i] = shape[i] - boxEnd[i];
      if (resultShape[i] <= 0) {
        resultShape[i] = 0;
        empty = True;
      }
    }
    return empty;
  }

} //# end namespace
