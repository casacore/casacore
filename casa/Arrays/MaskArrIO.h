//# MaskArrIO.h: Write out an ascii representation of a MaskedArray.
//# Copyright (C) 1993,1994,1995,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_MASKARRIO_H
#define CASA_MASKARRIO_H

#include <casa/aips.h>

//# Forward declarations
#include <casa/iosfwd.h>
namespace casa { //# NAMESPACE CASA - BEGIN

template<class T> class MaskedArray;
template<class T> class MaskedArray;


// <summary>
//    Ascii input/output operations for MaskedArrays.
// </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="tMaskArrIO">
//
// <prerequisite>
//   <li> <linkto class=Array>Array</linkto>
//   <li> <linkto class=MaskedArray>MaskedArray</linkto>
//   <li> <linkto group="ArrayIO.h#Array IO">ArrayIO</linkto>
// </prerequisite>
//
// <etymology>
// MaskArrIO is short for MaskedArrayIO, which is too long by the old
// AIPS++ file naming conventions.  This file contains global functions
// for writing out ascii representations of masked arrays.
// </etymology>
//
// <synopsis>
// These functions write out masked arrays in ascii representation.
// They simply write out the Array and the LogicalArray which is the mask
// using the ascii output functions for these objects.
// </synopsis>
//
// <example>
// <srcblock>
//   Vector<Int> a(10);
//   LogicalVector b(10);
//   MaskedArray m (a,b);
//      . . .
//   cout << m;
// </srcblock>
// This example writes out m in ascii.  It writes first a and then
// the mask constructed from b.
// </example>
//
// <motivation>
// These are primarily for debugging, so that one can examine the MaskedArray.
// Since MaskedArrays are manipulators of Arrays, it was not thought to
// be necessary, or even a good idea, to have other kinds of IO defined
// for them.
// </motivation>
//
// <linkfrom anchor="MaskedArray IO" classes="MaskedArray">
//    <here>MaskedArray IO</here> -- Ascii input/output operations
//    for MaskedArrays.
// </linkfrom>
//
// <group name="MaskedArray IO">


// 
// Write out an ascii representation of a MaskedArray.
// The component Array and LogicalArray are written out sequentially.
template<class T> ostream & operator<< (ostream &, const MaskedArray<T> &);


// </group>


} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <casa/Arrays/MaskArrIO.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
