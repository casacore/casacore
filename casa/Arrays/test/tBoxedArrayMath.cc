//# tBoxedArrayMath.cc: Test program for function boxedArrayMath
//# Copyright (C) 2008
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casacore/casa/Arrays/MaskArrMath.h>
#include <casacore/casa/Arrays/ArrayLogical.h>
#include <casacore/casa/Arrays/ArrayPartMath.h>
#include <casacore/casa/Arrays/ArrayIO.h>
#include <casacore/casa/Arrays/MaskArrIO.h>

#include <casacore/casa/Utilities/GenSort.h>
#include <casacore/casa/OS/Timer.h>
#include <iostream>


using namespace casacore;
using namespace std;

void doIt (Bool /*doTiming*/)
{
  {
    IPosition shape(2,5,5);
    Array<Float> arr(shape);
    indgen (arr);
    cout << boxedArrayMath(arr, IPosition(2,2), SumFunc<Float>()) << endl;
    cout << boxedArrayMath(arr, IPosition(2,1), SumFunc<Float>()) << endl;
    cout << boxedArrayMath(arr, IPosition(1,0), SumFunc<Float>()) << endl;

    cout << boxedArrayMath(arr, IPosition(4,2), SumFunc<Float>()) << endl;
    cout << boxedArrayMath(arr, IPosition(3,1), SumFunc<Float>()) << endl;
    cout << boxedArrayMath(arr, IPosition(),    SumFunc<Float>()) << endl;

    cout << boxedArrayMath(arr, IPosition(2,2), MedianFunc<Float>()) << endl;
    cout << boxedArrayMath(arr, IPosition(2,1), MedianFunc<Float>()) << endl;
    cout << boxedArrayMath(arr, IPosition(2,0), MedianFunc<Float>()) << endl;
  }

}

void doItMasked (Bool)
{
  {
    IPosition shape(2,5,5);
    Array<Float> darr(shape);
    indgen(darr);
    MaskedArray<Float> arr = darr(darr<=float(10) || darr>float(14));
    cout << arr.getMask() << endl;
    cout << sum(arr) << endl;
    cout << boxedArrayMath(arr, IPosition(2,2), MaskedSumFunc<Float>()) << endl;
    cout << boxedArrayMath(arr, IPosition(2,1), MaskedSumFunc<Float>()) << endl;

    cout << boxedArrayMath(arr, IPosition(4,2), MaskedSumFunc<Float>()) << endl;
    cout << boxedArrayMath(arr, IPosition(3,1), MaskedSumFunc<Float>()) << endl;

    cout << boxedArrayMath(arr, IPosition(2,2), MaskedMedianFunc<Float>()) << endl;
    cout << boxedArrayMath(arr, IPosition(2,1), MaskedMedianFunc<Float>()) << endl;
  }
}

int main(int argc, const char*[])
{
  try {
    doIt (argc>1);
    doItMasked (argc>1);
  } catch (exception& x) {
    cout << "Unexpected exception: " << x.what() << endl;
    return 1;
  }
  return 0;
}
