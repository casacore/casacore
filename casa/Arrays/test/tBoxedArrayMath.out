Axis Lengths: [3, 3]  (NB: Matrix in Row/Column order)
[12, 52, 41
 20, 60, 45
 13, 33, 24]

Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[0, 5, 10, 15, 20
 1, 6, 11, 16, 21
 2, 7, 12, 17, 22
 3, 8, 13, 18, 23
 4, 9, 14, 19, 24]

Axis Lengths: [1, 5]  (NB: Matrix in Row/Column order)
[10, 35, 60, 85, 110]

Axis Lengths: [3, 3]  (NB: Matrix in Row/Column order)
[12, 52, 41
 20, 60, 45
 13, 33, 24]

Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[0, 5, 10, 15, 20
 1, 6, 11, 16, 21
 2, 7, 12, 17, 22
 3, 8, 13, 18, 23
 4, 9, 14, 19, 24]

Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[0, 5, 10, 15, 20
 1, 6, 11, 16, 21
 2, 7, 12, 17, 22
 3, 8, 13, 18, 23
 4, 9, 14, 19, 24]

Axis Lengths: [3, 3]  (NB: Matrix in Row/Column order)
[3, 13, 20.5
 5, 15, 22.5
 6.5, 16.5, 24]

Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[0, 5, 10, 15, 20
 1, 6, 11, 16, 21
 2, 7, 12, 17, 22
 3, 8, 13, 18, 23
 4, 9, 14, 19, 24]

Axis Lengths: [1, 1]  (NB: Matrix in Row/Column order)
[12]

Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[1, 1, 1, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1]

250
Array: Axis Lengths: [3, 3]  (NB: Matrix in Row/Column order)
[12, 41, 41
 20, 35, 45
 13, 19, 24]

Mask:  Axis Lengths: [3, 3]  (NB: Matrix in Row/Column order)
[1, 1, 1
 1, 1, 1
 1, 1, 1]

Array: Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[0, 5, 10, 15, 20
 1, 6, 0, 16, 21
 2, 7, 0, 17, 22
 3, 8, 0, 18, 23
 4, 9, 0, 19, 24]

Mask:  Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[1, 1, 1, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1]

Array: Axis Lengths: [3, 3]  (NB: Matrix in Row/Column order)
[12, 41, 41
 20, 35, 45
 13, 19, 24]

Mask:  Axis Lengths: [3, 3]  (NB: Matrix in Row/Column order)
[1, 1, 1
 1, 1, 1
 1, 1, 1]

Array: Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[0, 5, 10, 15, 20
 1, 6, 0, 16, 21
 2, 7, 0, 17, 22
 3, 8, 0, 18, 23
 4, 9, 0, 19, 24]

Mask:  Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[1, 1, 1, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1]

Array: Axis Lengths: [3, 3]  (NB: Matrix in Row/Column order)
[3, 15, 20.5
 5, 17.5, 22.5
 6.5, 19, 24]

Mask:  Axis Lengths: [3, 3]  (NB: Matrix in Row/Column order)
[1, 1, 1
 1, 1, 1
 1, 1, 1]

Array: Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[0, 5, 10, 15, 20
 1, 6, 0, 16, 21
 2, 7, 0, 17, 22
 3, 8, 0, 18, 23
 4, 9, 0, 19, 24]

Mask:  Axis Lengths: [5, 5]  (NB: Matrix in Row/Column order)
[1, 1, 1, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1
 1, 1, 0, 1, 1]

